lang/LangPrimSource/SC_Clock.h

```diff
@@ -0,0 +1,269 @@
+		error("clock is not running.\n");
+		return errFailed;
+	}
+
+	double beats, seconds;
+	int err;
+
+	err = slotDoubleVal(b, &beats);
+	if (err) return errWrongType;
+
+	err = slotDoubleVal(&g->thread->seconds, &seconds);
+	if (err) return errWrongType;
+
+	clock->SetAll(clock->mTempo, beats, seconds);
+
+	return errNone;
+}
+
+template <typename Clock>
+int prClock_SetTempoAtBeat(struct VMGlobals *g, int numArgsPushed);
+template <typename Clock>
+int prClock_SetTempoAtBeat(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrSlot *a = g->sp - 2;
+	PyrSlot *b = g->sp - 1;
+	PyrSlot *c = g->sp;
+
+	Clock *clock = (Clock*)slotRawPtr(&slotRawObject(a)->slots[1]);
+	if (!clock) {
+		error("clock is not running.\n");
+		return errFailed;
+	}
+	if(clock->mTempo <= 0.f) {
+		error("cannot set tempo from this method. The message 'etempo_' can be used instead.\n");
+		// prTempoClock_SetTempoAtTime can be used.
+		return errFailed;
+	}
+
+	double tempo, beat;
+	int err = slotDoubleVal(b, &tempo);
+	if (err) return errFailed;
+	if (tempo <= 0.) {
+		error("invalid tempo %g. The message 'etempo_' can be used instead.\n", tempo);
+		// prTempoClock_SetTempoAtTime can be used.
+		return errFailed;
+	}
+
+	err = slotDoubleVal(c, &beat);
+	if (err) return errFailed;
+
+	clock->SetTempoAtBeat(tempo, beat);
+
+	return errNone;
+}
+
+template <typename Clock>
+int prClock_SetAll(struct VMGlobals *g, int numArgsPushed);
+template <typename Clock>
+int prClock_SetAll(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrSlot *a = g->sp - 3;
+	PyrSlot *b = g->sp - 2;
+	PyrSlot *c = g->sp - 1;
+	PyrSlot *d = g->sp;
+
+	Clock *clock = (Clock*)slotRawPtr(&slotRawObject(a)->slots[1]);
+	if (!clock) {
+		error("clock is not running.\n");
+		return errFailed;
+	}
+
+	double tempo, beat, secs;
+	int err = slotDoubleVal(b, &tempo);
+	if (err) return errFailed;
+
+	err = slotDoubleVal(c, &beat);
+	if (err) return errFailed;
+
+	err = slotDoubleVal(d, &secs);
+	if (err) return errFailed;
+
+	clock->SetAll(tempo, beat, secs);
+
+	return errNone;
+}
+
+template <typename Clock>
+int prClock_SetTempoAtTime(struct VMGlobals *g, int numArgsPushed);
+template <typename Clock>
+int prClock_SetTempoAtTime(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrSlot *a = g->sp - 2;
+	PyrSlot *b = g->sp - 1;
+	PyrSlot *c = g->sp;
+
+	Clock *clock = (Clock*)slotRawPtr(&slotRawObject(a)->slots[1]);
+	if (!clock) {
+		error("clock is not running.\n");
+		return errFailed;
+	}
+
+	double tempo, sec;
+	int err = slotDoubleVal(b, &tempo);
+	if (err) return errFailed;
+
+	err = slotDoubleVal(c, &sec);
+	if (err) return errFailed;
+
+	clock->SetTempoAtTime(tempo, sec);
+
+	return errNone;
+}
+
+
+template <typename DurationType>
+inline double DurToFloat(DurationType dur)
+{
+	using namespace std::chrono;
+	seconds secs         = duration_cast<seconds>(dur);
+	nanoseconds nanosecs = dur - secs;
+
+	return secs.count() + 1.0e-9 * nanosecs.count();
+}
+
+#ifdef SC_ABLETON_LINK
+void initLink();
+#endif
```

Comment:
I don't think there's any reason to declare this here, again because you've wrapped usage in guards

Reply:
`initLink` still needs to be declared before before being used in PyrSched.cpp. Maybe I should put its declaration just before its usage to avoid confusion ?

Reply:
In PyrSched.cpp:

```cpp
#ifdef SC_ABLETON_LINK
	initLink();
#endif
```

Reply:
ah sorry, never mind, I misread that as an `ifndef`